### PR TITLE
Handle cycles in type vars

### PR DESCRIPTION
**Pull request by flodiebold**
This might be the cause of #587.

### CODE REVIEW

File: crates/ra_hir/src/ty.rs [hunk 1]
```diff
@@ -1049,13 +1057,20 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
     /// Resolves the type completely; type variables without known type are
     /// replaced by Ty::Unknown.
-    fn resolve_ty_completely(&mut self, ty: Ty) -> Ty {
+    fn resolve_ty_completely(&mut self, tv_stack: &mut Vec<TypeVarId>, ty: Ty) -> Ty {
         ty.fold(&mut |ty| match ty {
             Ty::Infer(tv) => {
                 let inner = tv.to_inner();
+                if tv_stack.contains(&inner) {
+                    // recursive type
+                    return tv.fallback_value();
```
Comment (matklad) [inline crates/ra_hir/src/ty.rs]: that's a good place to place a mark!

File: crates/ra_hir/src/ty.rs [hunk 2]
```diff
@@ -1014,13 +1015,20 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
     /// be resolved as far as possible, i.e. contain no type variables with
     /// known type.
-    fn resolve_ty_as_possible(&mut self, ty: Ty) -> Ty {
+    fn resolve_ty_as_possible(&mut self, tv_stack: &mut Vec<TypeVarId>, ty: Ty) -> Ty {
         ty.fold(&mut |ty| match ty {
             Ty::Infer(tv) => {
                 let inner = tv.to_inner();
+                if tv_stack.contains(&inner) {
+                    // recursive type
+                    return tv.fallback_value();
```
Comment (matklad) [inline crates/ra_hir/src/ty.rs]: that's a good place to place a mark!

**Comment by flodiebold** [comment]
@matklad I added the marks :)

**Pull request merged**